### PR TITLE
Add Gibbs Sampler

### PR DESCRIPTION
**Pull request by kylie-box**
- Gibbs sampler has been tested, but it hasn't been fully integrated into the model pipeline yet. 
- One can get the model distribution instead of generating samples from the model distribution. 
- No adaptive softmax.

============================
Other changes have been made: 

- Learning rate scheduler
- gradient clipping
- gradient accumulation
- remove small cooccurrence counts for GPU sampler

### CODE REVIEW

File: hilbert/cooccurrence/cooccurrence_sector.py [hunk 1]
```diff
@@ -61,7 +61,7 @@ def __init__(
 
 
     @staticmethod
-    def load(path, sector, verbose=True):
+    def load(path, sector, rm_thres=None, verbose=True):
```
Comment (enewe101) [inline hilbert/cooccurrence/cooccurrence_sector.py]:
```suggestion
    def load(path, sector, min_cooccurrence_count=None, verbose=True):
```

Comment (enewe101) [reply hilbert/cooccurrence/cooccurrence_sector.py]: Great, glad this feature is making it in!  Can you rename as shown?

File: hilbert/cooccurrence/cooccurrence_sector.py [hunk 2]
```diff
@@ -119,19 +123,24 @@ def load_coo(cooccurrence_path, include_marginals=True, verbose=True):
...
+            Nx = sector._Nx.clone().detach().requires_grad_(True)
+            Nxt = sector._Nxt.clone().detach().requires_grad_(True)
+            # Nx = torch.tensor(sector._Nx, dtype=h.utils.get_dtype())
+            # Nxt = torch.tensor(sector._Nxt, dtype=h.utils.get_dtype())
             return data, I, J, Nx, Nxt
         else:
             return data, I, J
-
+    @staticmethod
+    def remove_small_numbers(cooc, threshold):
```
Comment (enewe101) [inline hilbert/cooccurrence/cooccurrence_sector.py]: Can you rename as follows:
```suggestion
    def remove_infrequent_cooccurrences(cooc, threshold):
```

File: hilbert/cooccurrence/cooccurrence_sector.py [hunk 3]
```diff
@@ -119,19 +123,24 @@ def load_coo(cooccurrence_path, include_marginals=True, verbose=True):
...
             I = torch.cat((I, add_i_idxs))
             J = torch.cat((J, add_j_idxs))
 
         if include_marginals:
             # Every sector has global marginals, so get marginals from last
             # sector.
-            Nx = torch.tensor(sector._Nx, dtype=h.utils.get_dtype())
-            Nxt = torch.tensor(sector._Nxt, dtype=h.utils.get_dtype())
+            Nx = sector._Nx.clone().detach().requires_grad_(True)
+            Nxt = sector._Nxt.clone().detach().requires_grad_(True)
```
Comment (enewe101) [inline hilbert/cooccurrence/cooccurrence_sector.py]: Lines 134, 135: why do we require grad?  We never backprop to data, so I think it should be `False`, no? 
```suggestion
            Nxt = sector._Nxt.clone().detach().requires_grad_(True)
```

Comment (enewe101) [reply hilbert/cooccurrence/cooccurrence_sector.py]: BTW, This is to get rid of the warning message right?

File: hilbert/factories.py [hunk 4]
```diff
@@ -88,18 +141,29 @@ def build_mle_sample_solver(
         temperature=2,            # MLE option
         batch_size=10000,
         balanced=False,
```
Comment (enewe101) [inline hilbert/factories.py]: Please add one thing, make balanced default to `True`.
```suggestion
         balanced=True,
```

File: hilbert/factories.py [hunk 5]
```diff
@@ -88,18 +141,29 @@ def build_mle_sample_solver(
...
+        gibbs=False,
         bias=False,
         init_embeddings_path=None,
         dimensions=300,
         learning_rate=0.01,
         opt_str='adam',
+        scheduler_str=None,
+        constant_fraction=1,
+        end_lr=0,
```
Comment (enewe101) [inline hilbert/factories.py]: Name should match `learning_rate` above
```suggestion
        end_learning_rate=0,
```

File: hilbert/factories.py [hunk 6]
```diff
@@ -88,18 +141,29 @@ def build_mle_sample_solver(
...
+        num_updates=1,
+        remove_threshold=None,
```
Comment (enewe101) [inline hilbert/factories.py]: Rename
```suggestion
        min_cooccurrence_count=None,
```

File: hilbert/factories.py [hunk 7]
```diff
@@ -123,22 +187,47 @@ def build_mle_sample_solver(
...
+    if gibbs:
+        print("Using Gibbs sampling.")
+        loader = h.loader.GibbsSampleLoader(
+            cooccurrence_path=cooccurrence_path,
+            learner=learner,
+            temperature=temperature,
+            batch_size=batch_size,
+            device=device,
+            verbose=verbose
+        )
```
Comment (enewe101) [inline hilbert/factories.py:206]: Should the `GibbsSampleLoader` also get passed in the low-cooccurrence count `remove_threshold` (or what I want to call the `min_cooccurrence_count`)?  Because the `CPUSampleLoader` and `GPUSampleLoader` both get it...

File: hilbert/factories.py [hunk 8]
```diff
@@ -123,22 +187,47 @@ def build_mle_sample_solver(
...
+            device=device,
+            verbose=verbose,
+            remove_threshold=remove_threshold,
+        )
+
 
     optimizer = get_optimizer(opt_str, learner, learning_rate)
 
+    if scheduler_str is not None:
+        lr_scheduler = get_lr_scheduler(scheduler_str=scheduler_str,
```
Comment (enewe101) [inline hilbert/factories.py]: I know that's a popular indent style, but in this code base we're always using just a four-space hanging indent style.
```suggestion
        lr_scheduler = get_lr_scheduler(
            scheduler_str=scheduler_str,
```

File: hilbert/solver.py [hunk 9]
```diff
@@ -77,24 +85,93 @@ def get_embeddings(self):
...
 
+                # self.V_gradient = list(self.learner.parameters())[0].grad
+                # self.W_gradient = list(self.learner.parameters())[1].grad
+                # self.W_gradient = self.learner.parameters().grad
+                # print("gradient of V: ", list(self.learner.parameters())[0].grad)
+                # print("gradient of W: ", list(self.learner.parameters())[1].grad)
+                # print("mean of V grad: ", torch.mean(self.V_gradient))
+                # print("mean of W grad: ", torch.mean(self.W_gradient))
+                # print("std of V grad: ", torch.std(self.V_gradient))
+                # print("std of W grad: ", torch.std(self.W_gradient))
```
Comment (enewe101) [inline hilbert/solver.py]: Remove these before we merge ok?

Comment (enewe101) [reply hilbert/solver.py]: "these" = commented out prints

File: hilbert/solver.py [hunk 10]
```diff
@@ -77,24 +85,93 @@ def get_embeddings(self):
...
+                                        "Positive sample word pairs are :{}\n"
+                                        "Negative sample word pairs are :{}\n".format(", ".join(map(str, pos_pairs)),
+                                                                                      ", ".join(map(str, neg_pairs))))
+                    except NameError:
+                        pass
+
                 # Take some steps
-                self.optimizer.step()
+                if (update_id + 1) % self.gradient_accumulation == 0\
+                        or (update_id + 1) == updates_per_cycle:
```
Comment (enewe101) [inline hilbert/solver.py:148]: Don't break out the `if` over multiple lines.  Assemble the full condition in a named boolean if needed.

E.g. 
```
    gradient_fully_accumulated = (update_id + 1) % self.gradient_accumulation == 0
    cycle_is_completing = (update_id + 1) == updates_per_cycle
    should_zero_grad = gradient_fully_accumulated or cycle_is_completing
    if should_zero_grad:
        ...
```

Comment (enewe101) [reply hilbert/solver.py:148]: Also, I want to understand better the thinking here, so let's look at this part of the code together ok?

File: hilbert/solver.py [hunk 11]
```diff
@@ -77,24 +85,93 @@ def get_embeddings(self):
     def get_params(self):
         return self.learner.get_params()
 
+    def get_batch_words(self, batch_id):
```
Comment (enewe101) [inline hilbert/solver.py:88]: My sense is that this function is for debugging only.  We can keep it if it is likely to be useful.  However, I think it should be on the loader itself, no?

File: hilbert/solver.py [hunk 12]
```diff
@@ -28,20 +30,26 @@ def __init__(
...
+        Solver.cycle() also support gradient accumulation calculation.
         """
 
         # Own it like you do
         self.loader = loader
         self.loss = loss
         self.optimizer = optimizer
         self.learner = learner
-        self.schedulers = schedulers or []
+        self.schedulers = schedulers
```
Comment (enewe101) [inline hilbert/solver.py]: I think we need to revert this.  If we don't have any schedulers, then it should be `[]` not None, since attempting to iterate over `None` will error.
```suggestion
        self.schedulers = schedulers or []
```

File: hilbert/factories.py [hunk 13]
```diff
@@ -88,18 +141,29 @@ def build_mle_sample_solver(
...
         batch_size=10000,
         balanced=False,
+        gibbs=False,
         bias=False,
         init_embeddings_path=None,
         dimensions=300,
         learning_rate=0.01,
         opt_str='adam',
+        scheduler_str=None,
+        constant_fraction=1,
```
Comment (enewe101) [inline hilbert/factories.py]: Just need some indication that this is related to the learning rate scheduler.
```suggestion
        lr_scheduler_constant_fraction=1,
```

File: hilbert/loader.py [hunk 14]
```diff
@@ -131,11 +131,12 @@ def __init__(
         temperature=1,
         batch_size=100000,
         device=None,
-        verbose=True
+        verbose=True,
+        remove_threshold=None,
```
Comment (enewe101) [inline hilbert/loader.py]: rename

File: hilbert/loader.py [hunk 15]
```diff
@@ -208,7 +209,241 @@ def describe(self):
...
+                                                                  steps=self.gibbs_iteration * 2,
+                                                                  get_distr=True)
+
+        return self.Pij, negative_sample_distrs
+
+    def sample(self, batch_size, toy=False):
+        '''
+
+        Gibbs sampler takes positive samples (i, j) drawn from corpus distribution Pij,
+        and sample a new j' from vocabulary by fixing i according to the conditional model distribution pi*e^(i dot j')
```
Comment (enewe101) [inline hilbert/loader.py]: This should say "by fixing i, and sampling j' according to the conditional model distribution, pj*e^(i dot j)"

File: hilbert/loader.py [hunk 16]
```diff
@@ -208,7 +209,241 @@ def describe(self):
...
+        # Randomly draw positive outcomes, and map them to ij pairs
+        positive_choices_idx = self.positive_sampler.sample(
+            sample_shape=(batch_size,))
+        # actual embeddings of choices
+        positive_samples = (self.I[positive_choices_idx].long(), self.J[positive_choices_idx].long())
+        positive_I, positive_J = positive_samples
+        IJ_sample = torch.empty(
+            (self.batch_size * 2, 2), device=self.device, dtype=torch.int64)
+
+        if toy:
```
Comment (enewe101) [inline hilbert/loader.py]: Please remove toy if we don't need it anymore (it's for debugging right?).  From what I understand it's duplicating the functionality of having `gibbs_iteration=1`

File: hilbert/loader.py [hunk 17]
```diff
@@ -208,7 +209,241 @@ def describe(self):
...
+            self,
+            cooccurrence_path,
+            learner,
+            temperature=1,
+            batch_size=1000,
+            gibbs_iteration=1,
+            get_distr=False,
+            device=None,
+            verbose=True
+            ):
```
Comment (enewe101) [inline hilbert/loader.py:225]: No `min_cooccurrences_count` argument (aka `remove_threshold`)?

File: hilbert/runners/run_base.py [hunk 18]
```diff
@@ -2,13 +2,20 @@
...
 from argparse import ArgumentParser
+import time
 
 
-def factory_args(args):
-    ignore = {
-        'save_embeddings_dir', 'num_writes', 'num_updates',
-        'monitor_closely', 'debug'
-    }
+def factory_args(args, keep_updates=True):
```
Comment (enewe101) [inline hilbert/runners/run_base.py:8]: Is there another way to do this?  Let's talk about it.

File: hilbert/runners/run_base.py [hunk 19]
```diff
@@ -51,11 +61,17 @@ def run(solver_factory, **args):
     # Train train train!  Write to disk once in awhile!
     updates_per_write = int(num_updates / num_writes)
     for write_num in range(num_writes):
-        solver.cycle(updates_per_write, monitor_closely)
+        try:
+            solver.cycle(updates_per_write, monitor_closely)
+        except h.exceptions.DivergenceError:
+            print("\n\nDied at {} epoch.".format(write_num))
+            tracer.declare(key='wrote_num', value=write_num+1)
+            raise h.exceptions.DivergenceError("Model has diverged")
```
Comment (enewe101) [inline hilbert/runners/run_base.py:68]: use either `print` or `tracer.declare`, not both!  Declare is better, because it prints and it logs to the trace.  You can bypass the key/value format for tracer if you just want to print anything, by using `tracer.trace()`.

File: hilbert/runners/run_base.py [hunk 20]
```diff
@@ -122,9 +159,44 @@ def add_batch_size_arg(parser):
...
+    parser.add_argument(
+        '--LR-scheduler', '-scheduler', default=None, choices=['linear', 'inverse', 'None'], dest='scheduler_str',
+        help="Type of learning rate scheduler"
+    )
+    parser.add_argument(
+        '--LR-scheduler-endLR', '-le', type=float, default=0.0, dest='end_lr',
+        help="The end learning rate for linear learning rate scheduler"
+    )
+    parser.add_argument(
+        '--constant-fraction', '-frac', type=float, default=0.1, dest='constant_fraction',
```
Comment (enewe101) [inline hilbert/runners/run_base.py]: Name this to make it more linked to lr_scheduler.  e.g. `--lr_scheduler_fraction`.

File: hilbert/scheduler.py [hunk 21]
```diff
@@ -31,23 +29,78 @@ def step(self):
...
+
+    def __init__(self, optimizer, start_lr, num_epochs, end_lr=0):
+        '''
+
+        :param optimizer: Optimizer used for training objective.
+        :param start_lr: Initial learning rate.
+        :param num_epochs: Number of epochs
+        :param end_lr: Stationary state learning rate
+        '''
+        super(LinearLRScheduler, self).__init__(optimizer,
```
Comment (enewe101) [inline hilbert/scheduler.py]: Minimal indent style
```suggestion
        super(LinearLRScheduler, self).__init__(
            optimizer,
```

File: hilbert/scheduler.py [hunk 22]
```diff
@@ -31,23 +29,78 @@ def step(self):
...
+        Based on the learning rate scheduler O(1/t) from Bergstra and Bengio (2012), where keeping learning rate
+        constant for the first [num_epochs] updates.
+        cur_lr = start_lr * num_epoch/max(t, num_epoch)
+
+        :param optimizer: Optimizer used for training objective.
+        :param start_LR: Initial learning rate.
+        :param num_epoch: Number of epochs keeping learning rate constant
+        """
+
+        super(InverseLRScheduler, self).__init__(optimizer,
```
Comment (enewe101) [inline hilbert/scheduler.py]: Minimal indent style
```suggestion
        super(InverseLRScheduler, self).__init__(
            optimizer,
```

File: hilbert/solver.py [hunk 23]
```diff
@@ -77,24 +85,93 @@ def get_embeddings(self):
...
+                    try:
+                        if self.cur_loss.item() > 1e4:
+                            # print("The last loss is: ", self.cur_loss)
+                            pos_pairs, neg_pairs = self.get_batch_words(batch_id)
+
+                            UserWarning("Extreme loss value is detected. current loss is greater than 1e4,\n"
+                                        "Positive sample word pairs are :{}\n"
+                                        "Negative sample word pairs are :{}\n".format(", ".join(map(str, pos_pairs)),
+                                                                                      ", ".join(map(str, neg_pairs))))
+                    except NameError:
```
Comment (enewe101) [inline hilbert/solver.py:142]: Is this `try...except` still needed?  (why?)

File: hilbert/tests/test_cooccurrence_sample_loader.py [hunk 24]
```diff
@@ -121,4 +121,5 @@ def test_cooccurrence_sample_loader_interface(self):
                 self.assertEqual(batch_id.dtype, torch.LongTensor.dtype)
         self.assertEqual(num_batches_seen, num_batches)
 
-
```
Comment (enewe101) [inline hilbert/tests/test_cooccurrence_sample_loader.py:124]: You can run tests when in `hilbert/hilbert` by doing:
```
    $ ./test [test_]file_name[.py] [Test]ClassName.[test_]function_name
```

**Review by enewe101 — CHANGES_REQUESTED** [review]
Hey Kylie,

So glad to see lots of great features coming into the main branch.  I've highlighted a few changes, and some areas where I want to sit together to discuss!  Have a look and then we'll sit down together!

e.

File: hilbert/solver.py (hunk 11 above)
Comment (kylie-box) [reply hilbert/solver.py:88]: I think it's because I'd like to get the actual words, instead of word ids. That requires the dictionary which we can't access from the loader no?

**Comment by enewe101** [comment]
Already added in using a different pull request
